Spam2 (Spam Management Dashboard)

keshavsethi opened the pull request:
Fixes #7927
**Refer https://github.com/publiclab/plots2/pull/7969#issuecomment-636247030 to see latest changes and UI**
and with the feedback from PR #7928
and in reference to #7961
@jywarren @SidharthBansal @ananyaarun @pydevsg @emilyashley @VladimirMikulic @cesswairimu @ebarry

I have made some changes as per feedback from @jywarren and others

1. **Made /spam2 as an alternate URL** so that it becomes easy for moderators and reviewers to see the changes and compare between two.
2. **Removed links and features that are not working** as suggested by @jywarren.now all the features in UI work except Dashboard which I made grey for reference.
3. **Added some Bulk moderation features** like bulk_spam, Bulk_publish, Select all/none, Bulk_delete, Bulk_ban.
4. **Made it responsive for all devices** and the table is scrollable horizontally within the card.
5. **Added color for status** of post as mentioned by @jywarren in #7953. I have commented there as well.
6. **New UI** which is simple and responsive. 
### Screenshots
![Screenshot from 2020-05-29 03-11-15](https://user-images.githubusercontent.com/36025262/83200088-2a7d2b80-a160-11ea-83a5-55e9f8fa4586.png)
![Screenshot from 2020-05-29 03-00-18](https://user-images.githubusercontent.com/36025262/83200095-2f41df80-a160-11ea-8320-5cbb46135f05.png)
![Screenshot from 2020-05-29 03-11-32](https://user-images.githubusercontent.com/36025262/83200106-34069380-a160-11ea-8b69-b230f38f39b9.png)
###Gifs
![ezgif com-video-to-gif (2)](https://user-images.githubusercontent.com/36025262/83200194-6fa15d80-a160-11ea-9e91-8274910e1612.gif)
![ezgif com-crop (2)](https://user-images.githubusercontent.com/36025262/83200197-716b2100-a160-11ea-8bad-ae860944339b.gif)

Tests are left as of now I am working on them. I will push them soon.
@jywarren @SidharthBansal @ananyaarun @pydevsg @emilyashley @VladimirMikulic @cesswairimu @ebarry Please give your feedbacks

Thanks!

keshavsethi commented:
> Items are not centred.
> 
> ![2020-05-29_00-38](https://user-images.githubusercontent.com/35997844/83200998-da44a000-a144-11ea-8c47-390a5602c8b9.png)

![Screenshot from 2020-05-29 07-12-06](https://user-images.githubusercontent.com/36025262/83259455-b2e9e380-a1d5-11ea-8465-d345d8ac5481.png)
I have aligned actions wrt other columns. 
Thanks!

cesswairimu commented:
Great job :tada: @keshavsethi :tada: , could you please take a look at the codeclimate issues and travis fails...ping me if you need any help

keshavsethi commented:
I have made a few more changes to UI and made it more refined and responsive. Please give your reviews. Please check bulk moderation features as well. Now I am working on some table related features and tests. 
![Screenshot from 2020-05-30 05-44-33](https://user-images.githubusercontent.com/36025262/83314804-dba3c480-a239-11ea-859f-54d9f171c79c.png)
![Screenshot from 2020-05-30 05-44-43](https://user-images.githubusercontent.com/36025262/83314808-de9eb500-a239-11ea-9364-670bfa8f2ada.png)
![Screenshot from 2020-05-30 05-45-37](https://user-images.githubusercontent.com/36025262/83314777-b6af5180-a239-11ea-9b9c-367ca6aff84c.png)
![Screenshot from 2020-05-30 05-45-46](https://user-images.githubusercontent.com/36025262/83314784-bdd65f80-a239-11ea-855a-2122d85e377c.png)

@jywarren @SidharthBansal @ananyaarun @pydevsg @emilyashley @VladimirMikulic @cesswairimu @ebarry Please give your feedback and reviews.
Thanks!! :smile:

VladimirMikulic commented on app/assets/stylesheets/spam2.css:
```diff
@@ -0,0 +1,124 @@
+  .nav_top  {
```
Please format this whole file.

keshavsethi replied:
Done!
Thanks!

VladimirMikulic replied:
There are still some serious formatting errors. Don't do this manually, use a tool.

keshavsethi replied:
I changed it and used a tool, is it fine now?

VladimirMikulic commented on app/views/spam2/_comments.html.erb:
```diff
@@ -0,0 +1,74 @@
+<script>
+$(document).ready(function() {
```
This block needs to be formatted and documented a bit.

keshavsethi replied:
I formatted it and added some comments

VladimirMikulic commented on app/views/spam2/_revisions.html.erb:
```diff
@@ -0,0 +1,68 @@
+<script>
+$(document).ready(function() {
```
If I am not mistaken we have this exact same block in the `app/views/spam2/_comments.html.erb` file. This needs to be refactored. If there is a bug in one block, we would need to copy-paste it into another block which is not very effective and can cause more even more bugs.

keshavsethi replied:
Both of them have different IDs and tables. 
` $("#comment_table").DataTable({` is in _comments.html.erb 
` $("#revision_table").DataTable({` is in _comments.html.erb

VladimirMikulic replied:
You could create a function to encapsulate the block and pass an ID as an argument, right?

keshavsethi replied:
Done!

VladimirMikulic commented on app/views/spam2/_revisions.html.erb:
```diff
@@ -0,0 +1,68 @@
+    });
+</script>
+
+<div class="card" id="table-card">
```
This Ruby block is also very similar to the one from `app/views/spam2/_comments.html.erb`. Refactoring this wouldn't hurt. @Uzay-G, what do you think?

keshavsethi replied:
I added it as function in spam2.js

VladimirMikulic commented on app/views/spam2/_spam.html.erb:
```diff
@@ -0,0 +1,109 @@
+<script>
+$(document).ready(function() {
```
Please keep the style consistent and format this block & add some comments.

keshavsethi replied:
Formatted and added a few comments
Thanks!

keshavsethi commented:
I have added spammed and unmoderated filter as suggested by @jywarren in #7953  and added count of them as well. 

![Screenshot from 2020-06-01 05-14-28](https://user-images.githubusercontent.com/36025262/83365721-eee49a80-a3c7-11ea-8820-9717c22d79f9.png)

Now I am adding a few more filters in card header. Most of Bulk moderation features are added. If more bulk moderation functions are required then please suggest. 
Thanks!! :smile:

Uzay-G commented on app/controllers/spam2_controller.rb:
```diff
@@ -0,0 +1,132 @@
+  def _spam
+    if logged_in_as(%w(moderator admin))
+      @nodes = Node.paginate(page: params[:page])
+      @nodes = if params[:type] == 'wiki'
```
It might be a bit more efficient and clearer here to first get the specific nodes and then paginate them.

keshavsethi replied:
Done, thanks!

Uzay-G commented on app/controllers/spam2_controller.rb:
```diff
@@ -0,0 +1,132 @@
+  end
+
+  def _spam_revisions
+    if logged_in_as(%w(amdin moderator))
```
please change `amdin` to `admin` here. Thanks :+1:

keshavsethi replied:
Sorry for that, Changed it!

Uzay-G commented on app/controllers/spam2_controller.rb:
```diff
@@ -0,0 +1,132 @@
+                            .order('timestamp DESC')
+      render template: 'spam2/_spam'
+    else
+      flash[:error] = 'Only moderators adn admins can moderate this.'
```
adn -> and

keshavsethi replied:
Changed it :sweat_smile:

Uzay-G commented on test/functional/spam2_controller_test.rb:
```diff
@@ -0,0 +1,5 @@
+require 'test_helper'
+
+class Spam2ControllerTest < ActionDispatch::IntegrationTest
+
```
Before we merge this large pr, we need to implement testing for the different routes here. I think this is essential with functional tests for the batch actions and then possibly a system test for the dashboard.

pydevsg replied:
Yes, rightly said @Uzay-G . I was going to write this. Thanks for bringing the point.

keshavsethi replied:
Sure, I am working on that. Thanks!!

keshavsethi replied:
Functional tests are done, please review. Thanks!!

cesswairimu commented:
Great great job @keshavsethi :rocket: @VladimirMikulic could you please do a final review against the changes you requested..Thanks

keshavsethi commented:
I have added some filters for notes, pages, spammed, and unmoderated posts. Their count is also displayed. Please refer to this gif. 
![filters](https://user-images.githubusercontent.com/36025262/83542111-2a8e7a00-a518-11ea-9ef2-379b77216c95.gif)
Please review this. Thanks!!

keshavsethi commented:
I have added modal for note body as sometimes it becomes big and doesn't fit in one column. I have also added a hide feature where users can hide some of its columns for a better view. 
**Gif**
![ezgif com-video-to-gif (3)](https://user-images.githubusercontent.com/36025262/83567984-19596380-a540-11ea-96b0-b253c76613c8.gif)
@jywarren @cesswairimu @VladimirMikulic @pydevsg @ananyaarun @Uzay-G Please review some latest changes, Thanks!!

jywarren commented:
Wow, this is really impressive work. I'll need to go through it carefully and hope to today. Thank you!

keshavsethi commented:
> Wow, this is really impressive work. I'll need to go through it carefully and hope to today. Thank you!

Thanks @jywarren, If you have any suggestions regarding some **extra filters** and **more bulk moderation features** then I can add them too. Can you please ask other PL moderators regarding some **special filter** or something else which they need in the dashboard.
Thank you!! :smile: :v:

emilyashley commented:
Thanks Keshav! I'm also putting 👀 on this this week :) The interface is looking great and I can't wait to pull down a branch and check it out in real time.

keshavsethi commented:
@cesswairimu @jywarren @emilyashley @VladimirMikulic  Can we merge this??

jywarren commented:
@emilyashley if this does not make any changes to the existing system (either back-end or UI), we could merge it as a beta, since it's at /spam2. 

I'm afraid I haven't had time to go through it thoroughly. @keshavsethi are there system tests for the UI so I can know exactly what to try out? Thank you!

keshavsethi commented:
> @emilyashley if this does not make any changes to the existing system (either back-end or UI), we could merge it as a beta, since it's at /spam2.
> 
> I'm afraid I haven't had time to go through it thoroughly. @keshavsethi are there system tests for the UI so I can know exactly what to try out? Thank you!

Right now this PR only has functional tests, I was thinking of adding system tests and other features in other PR as this PR is getting bigger. _If you want I can do system tests here only._
Thanks!!

keshavsethi commented:
@emilyashley @jywarren @cesswairimu @pydevsg @ananyaarun @VladimirMikulic  If you want me to work on /spam only then I can shift to that too. This PR is big because every page which was in /spam has to be here in /spam2 as well like wiki, notes, comments, and revisions. /spam2 has all the functionality as /spam including some of the bulk moderation features. If this PR is creating some issues like so big or something then I can either make PR in small parts or I can do this directly in /spam (This will not take much time). Please give your suggestions.
Thanks!!

SidharthBansal commented:
Spam2 is ambiguous to me. Kindly write some descriptive name for added functionality.

keshavsethi commented:
> Spam2 is ambiguous to me. Kindly write some descriptive name for added functionality.

@SidharthBansal 
1. Bulk moderation features in the second top nav. their methods are named as batch_x example batch_spam, batch_publish etc, and ids are named as batch-x like batch-spam etc.
2. Datatables Ids are named as x_table eg. node_table, revision_table etc. It has sort and search filters.
3. Table filter, I have added some basic table filters like spammed for filtering spam post and so on. their js code is in _node.html.erb.
4. Node individual publish and spamming. for this alter is added and modal is created for better moderation. It is in the table of _node.html.erb.
5. Overall UI, you can find most of it like top nav and sidebar in spam.html.erb and admin.css

Thanks!!

jywarren commented:
Hi, thank you! So glad to see the /spam2 URL and will review and hopefully merge today. This will be great. Confirming that nothing in /spam has been changed and will work the same?

jywarren commented:
Yay!!!

And sorry to repeat myself on system tests. Let's do them in future PRs! If u can open a new PR with system tests for this new code, that'd be great, and link it to here. 🎉 

Especially system tests for bulk moderation tasks, as it can happen that future code breaks such carefully written UI and we want to ensure all the JS is working. Thank you!!!

One small UI change I might request is to use a button group for the right-side buttons, which can also make it more clear that the grey ones are not enabled, you know? They could be quite small: https://getbootstrap.com/docs/4.4/components/button-group/#sizing

keshavsethi commented:
> Hi, thank you! So glad to see the /spam2 URL and will review and hopefully merge today. This will be great. Confirming that nothing in /spam has been changed and will work the same?

Yes, Nothing has been changed by me in /spam and it will work the same

keshavsethi commented:
> I love this. Very very thorough functional testing. I think this is good to go! Can I request that this UI be system tested thoroughly as well? For future PRs, having system tests means we can read through the specific UI interactions and be sure we understand them as well, so if you could think ahead to including them that would be great and would help speed up reviews too!
> 
> Example of a system test here: https://github.com/publiclab/plots2/blob/master/test/system/post_question_test.rb
> 
> Great job on this @keshavsethi!!!

@jywarren I am working on system tests and I will make a PR for this soon.
Thanks!!

keshavsethi commented:
> Yay!!!
> 
> And sorry to repeat myself on system tests. Let's do them in future PRs! If u can open a new PR with system tests for this new code, that'd be great, and link it to here. 
> 
> Especially system tests for bulk moderation tasks, as it can happen that future code breaks such carefully written UI and we want to ensure all the JS is working. Thank you!!!
> 
> One small UI change I might request is to use a button group for the right-side buttons, which can also make it more clear that the grey ones are not enabled, you know? They could be quite small: https://getbootstrap.com/docs/4.4/components/button-group/#sizing

@jywarren Sure, I will add a button group for this. Thanks!!

cesswairimu commented:
:tada:
